File: samples/fpc/page_cache.pas
{
  page_cache.pas — one shared cache vs two private caches.

  The fbintf twin of ../cpp/page_cache.cpp: the same hot-page ping-pong
  run against both cache topologies, refereed by per-attachment
  MON$IO_STATS:

    phase 1 — two client processes -> ONE SuperServer shared cache
              (coherency by shared memory; almost no physical I/O)
    phase 2 — two EMBEDDED engine processes with PRIVATE caches over one
              file (ServerMode=SuperClassic sandbox; coherency by LCK_bdb
              page locks + blocking ASTs — data travels through the disk,
              so the same workload turns into hundreds of reads AND writes)

  The parent only spawns children (TProcess re-invoking this binary with a
  role argument — the Pascal spelling of the C++ twin's fork/exec); every
  attachment lives in a child process, and each embedded child IS a full
  engine because its connection string has no host part.  Rows 1 and 2
  share a data page, so the two writers fight over one page without ever
  touching one row.  See ../../page-cache-coherency.md.

  Build & run (see ../README.md):
      make -C samples/fpc bin/page_cache && samples/fpc/bin/page_cache
}
program page_cache;

{$mode delphi}
{$codepage UTF8}
{$interfaces COM}

uses SysUtils, Classes, Process, BaseUnix, IB, FBHandsOn;

const
  SRV_DB = 'localhost:/tmp/fbhandson/page_cache_srv_fpc.fdb';
  EMB_DB = '/tmp/fbhandson/page_cache_emb_fpc.fdb';
  SANDBOX = '/tmp/fbhandson/fbemb_fpc';
  ROUNDS = 300;

var UseSandboxEnv: boolean = false;   { phase 2: children get FIREBIRD=SANDBOX }

procedure InitDb(const conn: AnsiString);      { --init: fresh table, two rows }
var A: IAttachment;
    T: ITransaction;
begin
  A := AttachOrCreate(conn);
  T := A.StartTransaction([isc_tpb_write, isc_tpb_nowait, isc_tpb_concurrency],
    taCommit);
  try A.ExecuteSQL(T, 'drop table t', []);
  except on EIBInterBaseError do ; end;
  T.Commit;
  T := A.StartTransaction([isc_tpb_write, isc_tpb_nowait, isc_tpb_concurrency],
    taCommit);
  A.ExecuteSQL(T, 'create table t (id int primary key, v int)', []);
  T.Commit;
  T := A.StartTransaction([isc_tpb_write, isc_tpb_nowait, isc_tpb_concurrency],
    taCommit);
  A.ExecuteSQL(T, 'insert into t values (1, 0)', []);
  A.ExecuteSQL(T, 'insert into t values (2, 0)', []);
  T.Commit;
end;

procedure Worker(const conn, rowId: AnsiString);              { --worker }
var A: IAttachment;
    T: ITransaction;
    R: IResultset;
    i: integer;
begin
  A := FirebirdAPI.OpenDatabase(conn, DefaultDPB);
  for i := 1 to ROUNDS do
  begin
    T := A.StartTransaction([isc_tpb_write, isc_tpb_wait, isc_tpb_concurrency],
      taCommit);
    A.ExecuteSQL(T, 'update t set v = v + 1 where id = ' + rowId, []);
    T.Commit;
  end;
  T := A.StartTransaction([isc_tpb_read, isc_tpb_nowait, isc_tpb_concurrency],
    taCommit);
  R := A.OpenCursorAtStart(T,
    'select MON$PAGE_FETCHES, MON$PAGE_READS, MON$PAGE_WRITES ' +
    'from MON$IO_STATS join MON$ATTACHMENTS using (MON$STAT_ID) ' +
    'where MON$ATTACHMENT_ID = CURRENT_CONNECTION');
  writeln(Format('  worker pid %-6d row %s: %d commits | page fetches=%-6s ' +
    'reads=%-4s writes=%s', [FpGetPid, rowId, ROUNDS,
    R[0].AsString, R[1].AsString, R[2].AsString]));
  T.Commit;
end;

procedure Check(const conn: AnsiString);       { --check: any lost updates? }
var A: IAttachment;
    T: ITransaction;
    R: IResultset;
begin
  A := FirebirdAPI.OpenDatabase(conn, DefaultDPB);
  T := A.StartTransaction([isc_tpb_read, isc_tpb_nowait, isc_tpb_concurrency],
    taCommit);
  R := A.OpenCursorAtStart(T, 'select id, v from t order by id');
  while not R.IsEof do
  begin
    writeln(Format('  final: id=%s v=%s (expected %d)',
      [R[0].AsString, R[1].AsString, ROUNDS]));
    R.FetchNext;
  end;
  T.Commit;
end;

function Spawn(const args: array of AnsiString): TProcess;
var i: integer;
begin
  Result := TProcess.Create(nil);
  Result.Executable := ParamStr(0);
  for i := 0 to High(args) do
    Result.Parameters.Add(args[i]);
  if UseSandboxEnv then
  begin  { children inherit our environment plus the sandbox FIREBIRD root }
    for i := 1 to GetEnvironmentVariableCount do
      Result.Environment.Add(GetEnvironmentString(i));
    Result.Environment.Add('FIREBIRD=' + SANDBOX);
  end;
  Result.Execute;
end;

procedure Run1(const args: array of AnsiString);   { spawn and await one }
var P: TProcess;
begin
  Flush(Output);
  P := Spawn(args);
  P.WaitOnExit;
  P.Free;
end;

procedure RunWorkers(const conn: AnsiString);      { two writers in parallel }
var P1, P2: TProcess;
begin
  Flush(Output);
  P1 := Spawn(['--worker', conn, '1']);
  P2 := Spawn(['--worker', conn, '2']);
  P1.WaitOnExit;
  P2.WaitOnExit;
  P1.Free;
  P2.Free;
end;

{ Build the embedded sandbox: a FIREBIRD root whose firebird.conf says
  SuperClassic, so each process locks the file SHARED and runs its own
  page cache (see 'Three layers of arbitration'). }
procedure MakeSandbox;
const links: array[0..4] of AnsiString = ('plugins', 'intl', 'tzdata',
    'firebird.msg', 'security6.fdb');
var i: integer;
    conf: TextFile;
begin
  FpMkdir(SANDBOX, &777);
  for i := 0 to High(links) do
    FpSymlink(PAnsiChar('/opt/firebird/' + links[i]),
      PAnsiChar(SANDBOX + '/' + links[i]));
  AssignFile(conf, SANDBOX + '/firebird.conf');
  Rewrite(conf);
  writeln(conf, 'ServerMode = SuperClassic');
  CloseFile(conf);
end;

begin
  if (ParamCount >= 2) and (ParamStr(1) = '--init') then
    begin InitDb(ParamStr(2)); exit; end;
  if (ParamCount >= 3) and (ParamStr(1) = '--worker') then
    begin Worker(ParamStr(2), ParamStr(3)); exit; end;
  if (ParamCount >= 2) and (ParamStr(1) = '--check') then
    begin Check(ParamStr(2)); exit; end;

  writeln('phase 1: two client processes, ONE SuperServer shared cache');
  Run1(['--init', SRV_DB]);
  RunWorkers(SRV_DB);
  Run1(['--check', SRV_DB]);

  MakeSandbox;
  DeleteFile(EMB_DB);

  writeln('phase 2: two EMBEDDED engine processes, PRIVATE page caches');
  UseSandboxEnv := true;
  Run1(['--init', EMB_DB]);
  RunWorkers(EMB_DB);
  Run1(['--check', EMB_DB]);
  writeln('same workload - the private caches paid for coherency in disk I/O.');
end.

File: samples/fb-cpp/page_cache.cpp
/*
 *  page_cache.cpp (fb-cpp) — one shared cache vs two private caches.
 *
 *  The fb-cpp twin of ../cpp/page_cache.cpp: the same hot-page ping-pong
 *  run against both cache topologies (phase 1: two clients of one
 *  SuperServer shared cache; phase 2: two EMBEDDED engine processes with
 *  private caches over one file, via a SuperClassic sandbox this program
 *  builds itself), refereed by per-attachment MON$IO_STATS.  fb-cpp fetches
 *  each worker's I/O counters straight into a struct with
 *  queryFirstRowAs<IoStats>(); the fork/exec choreography is identical —
 *  cache topology is decided by the connection string, not the wrapper.
 *  See ../../page-cache-coherency.md.
 *
 *  Build & run (see ../README.md):
 *      ./build/fbcpp_page_cache
 */

#include "fbcpp_sample.h"
#include <cstdio>
#include <cstring>
#include <string>
#include <vector>
#include <sys/stat.h>
#include <sys/wait.h>
#include <unistd.h>

using namespace fbcpp;
using namespace fbcpp_sample;

static const char* SRV_DB = "inet://localhost//tmp/fbhandson/page_cache_srv_fbcpp.fdb";
static const char* EMB_DB = "/tmp/fbhandson/page_cache_emb_fbcpp.fdb";
static const char* SANDBOX = "/tmp/fbhandson/fbemb_fbcpp";
static const int ROUNDS = 300;

struct IoStats
{
	std::int64_t fetches, reads, writes;
};

static int initDb(const char* conn)          // --init: fresh table, two rows
{
	Client client{"fbclient"};
	Attachment att = attachOrCreate(client, conn);
	{
		Transaction t{att};
		try { att.execute(t, "drop table t"); }
		catch (const DatabaseException&) {}
		t.commit();
	}
	Transaction ddl{att};
	att.execute(ddl, "create table t (id int primary key, v int)");
	ddl.commit();
	Transaction ins{att};
	att.execute(ins, "insert into t values (1, 0)");
	att.execute(ins, "insert into t values (2, 0)");
	ins.commit();
	return 0;
}

static int worker(const char* conn, const char* rowId)   // --worker
{
	Client client{"fbclient"};
	Attachment att{client, conn, defaultOptions()};
	const std::string update = std::string("update t set v = v + 1 where id = ") + rowId;
	for (int i = 0; i < ROUNDS; ++i)
	{
		Transaction t{att};
		att.execute(t, update);
		t.commit();
	}
	Transaction t{att};
	const IoStats io = att.queryFirstRowAs<IoStats>(t,
		"select MON$PAGE_FETCHES, MON$PAGE_READS, MON$PAGE_WRITES "
		"from MON$IO_STATS join MON$ATTACHMENTS using (MON$STAT_ID) "
		"where MON$ATTACHMENT_ID = CURRENT_CONNECTION").value();
	printf("  worker pid %-6d row %s: %d commits | page fetches=%-6lld "
		"reads=%-4lld writes=%lld\n", getpid(), rowId, ROUNDS,
		(long long) io.fetches, (long long) io.reads, (long long) io.writes);
	t.commit();
	return 0;
}

static int check(const char* conn)           // --check: any lost updates?
{
	Client client{"fbclient"};
	Attachment att{client, conn, defaultOptions()};
	Transaction t{att};
	Statement stmt{att, t, "select id, v from t order by id"};
	for (bool ok = stmt.execute(t); ok; ok = stmt.fetchNext())
		printf("  final: id=%d v=%d (expected %d)\n",
			stmt.getInt32(0).value_or(-1), stmt.getInt32(1).value_or(-1), ROUNDS);
	t.commit();
	return 0;
}

static void run(const char* self, std::initializer_list<const char*> args)
{
	if (fork() == 0)
	{
		std::vector<const char*> v{self};
		v.insert(v.end(), args.begin(), args.end());
		v.push_back(nullptr);
		execv(self, const_cast<char**>(v.data()));
		_exit(127);
	}
}

static void await() { while (wait(nullptr) > 0) {} }

int main(int argc, char** argv)
{
	try
	{
		if (argc > 2 && !strcmp(argv[1], "--init"))   return initDb(argv[2]);
		if (argc > 3 && !strcmp(argv[1], "--worker")) return worker(argv[2], argv[3]);
		if (argc > 2 && !strcmp(argv[1], "--check"))  return check(argv[2]);

		printf("phase 1: two client processes, ONE SuperServer shared cache\n");
		fflush(stdout);
		run(argv[0], {"--init", SRV_DB}); await();
		run(argv[0], {"--worker", SRV_DB, "1"});
		run(argv[0], {"--worker", SRV_DB, "2"});
		await();
		run(argv[0], {"--check", SRV_DB}); await();

		// Build the embedded sandbox: a FIREBIRD root whose firebird.conf
		// says SuperClassic, so each process locks the file SHARED and runs
		// its own page cache (see 'Three layers of arbitration').
		mkdir(SANDBOX, 0777);
		for (const char* f : {"plugins", "intl", "tzdata", "firebird.msg", "security6.fdb"})
			symlink((std::string("/opt/firebird/") + f).c_str(),
				(std::string(SANDBOX) + "/" + f).c_str());
		FILE* conf = fopen((std::string(SANDBOX) + "/firebird.conf").c_str(), "w");
		fputs("ServerMode = SuperClassic\n", conf);
		fclose(conf);
		setenv("FIREBIRD", SANDBOX, 1);
		unlink(EMB_DB);

		printf("phase 2: two EMBEDDED engine processes, PRIVATE page caches\n");
		fflush(stdout);
		run(argv[0], {"--init", EMB_DB}); await();
		run(argv[0], {"--worker", EMB_DB, "1"});
		run(argv[0], {"--worker", EMB_DB, "2"});
		await();
		run(argv[0], {"--check", EMB_DB}); await();
		printf("same workload — the private caches paid for coherency in disk I/O.\n");
		return 0;
	}
	catch (const std::exception& e)
	{
		return report(e);
	}
}
